packages/jest-circus/src/utils.js

```diff
@@ -0,0 +1,225 @@
+      case 'afterAll':
+        result.afterAll.push(hook);
+        break;
+    }
+  }
+
+  return result;
+};
+
+const getEachHooksForTest = (
+  test: Test,
+): {[key: 'beforeEach' | 'afterEach']: Array<Hook>} => {
+  const result = {afterEach: [], beforeEach: []};
+  let {parent: block} = test;
+
+  do {
+    for (const hook of block.hooks) {
+      switch (hook.type) {
+        case 'beforeEach':
+          // Before hooks are executed from top to bottom, the opposite of the
+          // way we traversed it.
+          result.beforeEach.unshift(hook);
+          break;
+        case 'afterEach':
+          result.afterEach.push(hook);
+          break;
+      }
+    }
+  } while ((block = block.parent));
+  return result;
+};
+
+const callAsyncFn = (
+  fn: AsyncFn,
+  testContext: ?TestContext,
+  {isHook}: {isHook?: boolean} = {isHook: false},
+): Promise<any> => {
+  // If this fn accepts `done` callback we return a promise that fullfills as
+  // soon as `done` called.
+  if (fn.length) {
+    return new Promise((resolve, reject) => {
+      const done = (reason?: Error | string): void =>
+        reason ? reject(reason) : resolve();
+
+      fn.call(testContext, done);
+    });
+  }
+
+  let returnedValue;
+  try {
+    returnedValue = fn.call(testContext);
+  } catch (error) {
+    return Promise.reject(error);
+  }
+
+  // If it's a Promise, return it.
+  if (returnedValue instanceof Promise) {
+    return returnedValue;
+  }
+
+  if (!isHook && returnedValue !== void 0) {
+    throw new Error(
+      `
+      test functions can only return Promise or undefined.
+      Returned value: ${String(returnedValue)}
+      `,
+    );
+  }
+
+  // Otherwise this test is synchronous, and if it didn't throw it means
+  // it passed.
+  return Promise.resolve();
+};
+
+const getTestDuration = (test: Test): ?number => {
+  const {startedAt} = test;
+  return startedAt ? Date.now() - startedAt : null;
+};
+
+const makeTestResults = (describeBlock: DescribeBlock): TestResults => {
+  let testResults = [];
+  for (const test of describeBlock.tests) {
+    const testPath = [];
+    let parent = test;
+    do {
+      testPath.unshift(parent.name);
+    } while ((parent = parent.parent));
+
+    const {status} = test;
+
+    if (!status) {
+      console.log(test);
+      throw new Error('status should be present after tests are run');
+    }
+    testResults.push({
+      duration: test.duration,
+      errors: test.errors.map(_formatError),
+      status,
+      testPath,
+    });
+  }
+
+  for (const child of describeBlock.children) {
+    testResults = testResults.concat(makeTestResults(child));
+  }
+
+  return testResults;
+};
+
+// Return a string that identifies the test (concat of parent describe block
+// names + test title)
+const getTestID = (test: Test) => {
+  const titles = [];
+  let parent = test;
+  do {
+    titles.unshift(parent.name);
+  } while ((parent = parent.parent));
+
+  titles.shift(); // remove TOP_DESCRIBE_BLOCK_NAME
```

Comment:
we do. In order to use `beforeEach` or `test` at the top level we need to have a virtual top level `describe` that will wrap everything. But we also want to make it invisible for the outside world